### ISSUE TITLE
Inconsistent behaviour for `utcOffset`.

### ISSUE THREAD

**Issue by y-abarnikov** [iamkun/dayjs#1340]
**Describe the bug**
Inconsistent behaviour for `utcOffset`. Probably issue with Daylight Saving Time.
![image](https://user-images.githubusercontent.com/40033983/104925721-dac9c980-59a7-11eb-87b4-4ee00024bb14.png)

**Expected behavior**
Should be `-4` as in the row below.

**Information**
 - Day.js Version [1.10.3]
 - moment-timezone Version [0.5.32]
 - OS: [Mac OS Big Sur]
 - Interpreter Node.js [14.15.0]
 - Time zone: [GMT+2 (Eastern European Time)]

**Comment by y-abarnikov** [comment]
@iamkun

**Comment by addisonElliott** [comment]
FYI, it seems the bug is in moment to me.

Right now the correct offset for `America/New_York` is -5.

See here: https://www.zeitverschiebung.net/en/timezone/america--new_york

**Comment by hansdaniels** [comment]
DST is picked up by `tz`, but strangely, it is added to the hour only, not to the offset:
```javascript
import dayjs from 'dayjs';
import tz from 'dayjs/plugin/timezone.js';
import utc from 'dayjs/plugin/utc.js';
dayjs.extend(tz);
dayjs.extend(utc);

console.log(dayjs.tz(new Date('2020-03-29T00:30:00Z').getTime(), 'Europe/Berlin').format(), 'is correct');
console.log(dayjs.tz(new Date('2020-03-29T01:30:00Z').getTime(), 'Europe/Berlin').format(), 'should be 2020-03-29T03:30:00+02:00, because DST kicked in half an hour before');
```
yields:
```
console.log(dayjs.tz(new Date('2020-03-29T00:30:00Z').getTime(), 'Europe/Berlin').format(), 'is correct');
console.log(dayjs.tz(new Date('2020-03-29T01:30:00Z').getTime(), 'Europe/Berlin').format(), 'should be 2020-03-29T03:30:00+02:00, because DST kicked in half an hour before');
```

**Comment by addisonElliott** [comment]
Not sure your bug is similar to the OPs, unless I'm missing the connection.

But I did confirm and figure out what the problem is with your case. It's actually a pretty easy fix (I think).

## The problem

I'm in CST, which has an offset of -5/-6 (depending on DST). Let's use this as an example.

Given the input you gave: `2020-03-29T00:30:00Z`

The code converts to local machines timezone (**at this time it was -5 for me**). Then it converts to the desired timezone. These are the results:
    * `2020-03-28T20:30:00-05:00`
    * `2020-03-29T03:30:00+02:00`

Then these two dates are subtracted (disregard the offset), giving -07:00. Next, we take local machine offset (-5) and subtract from -7. -5 - (-7) = +2. Correct answer.

That's how it **should** work. Let's look at the code and I'll show you the bug.

```javascript
  proto.tz = function (timezone = defaultTimezone, keepLocalTime) {
    const oldOffset = this.utcOffset()
    const target = this.toDate().toLocaleString('en-US', { timeZone: timezone })
    const diff = Math.round((this.toDate() - new Date(target)) / 1000 / 60)
    let ins = d(target).$set(MS, this.$ms).utcOffset(localUtcOffset - diff, true)
    if (keepLocalTime) {
      const newOffset = ins.utcOffset()
      ins = ins.add(oldOffset - newOffset, MIN)
    }
    ins.$x.$timezone = timezone
    return ins
  }
```
Source: https://github.com/iamkun/dayjs/blob/dev/src/plugin/timezone/index.js#L101

Check out that it takes `localUtcOffset` and subtracts the diff. That `localUtcOffset` is obtained by doing the following. The problem is that DST is applied now so the offset for me is -06:00, giving the offset of +01:00 that we're seeing.

```javascript
  const localUtcOffset = d().utcOffset()
```
Source: https://github.com/iamkun/dayjs/blob/dev/src/plugin/timezone/index.js#L39

## The fix

Replace `localUtcOffset` with `this.utcOffset()` and that will return the UTC offset **at that date**

## What you should do

I haven't tested this but I think it'll work.

If that works for you, I'd encourage you to submit a PR and fix this.

## When was the bug introduced

Bug has existed ever since the timezone plugin was added.

https://github.com/iamkun/dayjs/pull/974/files

**Comment by hansdaniels** [comment]
Thanks!
I'll do as you have suggested, but I won't find the time for it today.

Am 21. Januar 2021 18:38:56 MEZ schrieb Addison Elliott <notifications@github.com>:

**Comment by hansdaniels** [comment]
The proposed works for utcOffset when starting from `Z`.
But when the input date already has a different timezone, `tz` is now broken.

```
 FAIL  test/plugin/timezone.test.js (5.826s)
  ● Parse › parse and convert between timezones
    Expected value to be:
      "2014-06-01T09:00:00-07:00"
    Received:
      "2014-06-01T09:00:00-11:00"

      57 |   it('parse and convert between timezones', () => {
      58 |     const newYork = dayjs.tz('2014-06-01 12:00', NY)
    > 59 |     expect(newYork.tz('America/Los_Angeles').format()).toBe('2014-06-01T09:00:00-07:00')
      60 |     expect(newYork.tz('Europe/London').format()).toBe('2014-06-01T17:00:00+01:00')
```

Have you meant changing 
```javascript
    let ins = d(target).$set(MS, this.$ms).utcOffset(localUtcOffset - diff, true)
```
to
```javascript
let ins = d(target).$set(MS, this.$ms).utcOffset(this.utcOffset() - diff, true)
```
or have you had another line in mind?
This was the only reference to `localUtcOffset`.

Do you directly see the cause of this or would you have to dig deeper as well?

**Comment by addisonElliott** [comment]
No you changed the code correctly. Good catch.

I think I know the problem. Try the below snippet. Basically, the issue is that `.utcOffset` is overridden when a timezone is set, but we actually want the user's local timezone at that date. The reason is because `new Date` in Javascript without a timezone will assume local timezone and we need that offset.

Another thing is I'm not sure that `keepLocalTime` logic is valid anymore. Curious to see if tests fail for that too.

```javascript
  proto.tz = function (timezone = defaultTimezone, keepLocalTime) {
    const oldOffset = this.utcOffset()
    const date = this.toDate();
    const target = date.toLocaleString('en-US', { timeZone: timezone })
    const diff = Math.round((date - new Date(target)) / 1000 / 60)
    let ins = d(target).$set(MS, this.$ms).utcOffset((-Math.round(date.getTimezoneOffset() / 15) * 15) - diff, true)
    if (keepLocalTime) {
      const newOffset = ins.utcOffset()
      ins = ins.add(oldOffset - newOffset, MIN)
    }
    ins.$x.$timezone = timezone
    return ins
  }
```

**Comment by iamkun** [comment]
fixed in #1352
